contracts/asset-proxy/contracts/src/bridges/UniswapV2Bridge.sol

```diff
@@ -0,0 +1,140 @@
+import "../interfaces/IUniswapV2Router01.sol";
+import "../interfaces/IERC20Bridge.sol";
+
+
+// solhint-disable space-after-comma
+// solhint-disable not-rely-on-time
+contract UniswapV2Bridge is
+    IERC20Bridge,
+    IWallet,
+    DeploymentConstants
+{
+    using LibAddressArray for address[];
+
+    struct TransferState {
+        address fromTokenAddress;
+        uint256 fromTokenBalance;
+        uint256 boughtAmount;
+    }
+
+    /// @dev Callback for `IERC20Bridge`. Tries to buy `amount` of
+    ///      `toTokenAddress` tokens by selling the entirety of the `fromTokenAddress`
+    ///      token encoded in the bridge data.
+    /// @param toTokenAddress The token to buy and transfer to `to`.
+    /// @param from The maker (this contract).
+    /// @param to The recipient of the bought tokens.
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param bridgeData The abi-encoded "from" token address.
+    /// @return success The magic bytes if successful.
+    function bridgeTransferFrom(
+        address toTokenAddress,
+        address from,
+        address to,
+        uint256 amount,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        // hold variables to get around stack depth limitations
+        TransferState memory state;
+
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (state.fromTokenAddress) = abi.decode(bridgeData, (address));
+
+        // Just transfer the tokens if they're the same.
+        if (state.fromTokenAddress == toTokenAddress) { // NOT TESTED
+            LibERC20Token.transfer(state.fromTokenAddress, to, amount);
+            return BRIDGE_SUCCESS;
+        }
+
+        // Get our balance of `fromTokenAddress` token.
+        state.fromTokenBalance = IERC20Token(state.fromTokenAddress).balanceOf(address(this));
+
+        // Grant the Uniswap router an allowance.
+        LibERC20Token.approve(
+            state.fromTokenAddress,
+            _getUniswapV2Router01Address(),
+            state.fromTokenBalance
+        );
+
+        // Convert directly from fromTokenAddress to toTokenAddress
+        address[] memory path = new address[](2);
+        path[0] = state.fromTokenAddress;
+        path[1] = toTokenAddress;
+
+        // Buy as much `toTokenAddress` token with `fromTokenAddress` token
+        // and transfer it to `to`.
+        IUniswapV2Router01 router = IUniswapV2Router01(_getUniswapV2Router01Address());
+        uint[] memory amounts = router.swapExactTokensForTokens(
+             // Sell all tokens we hold.
+            state.fromTokenBalance,
+             // Minimum buy amount.
+            amount,
+            // Convert `fromTokenAddress` to `toTokenAddress`.
+            path,
+            // Recipient is `to`.
+            to,
+            // Expires after this block.
+            block.timestamp
+        );
+
+        state.boughtAmount = amounts[0];
```

Comment:
```

| amounts | uint[] memory | The input token amount and all subsequent output token amounts.|

```
https://uniswap.org/docs/v2/smart-contracts/router/#swapexacttokensfortokens

does this mean the first element is amountSold?

Reply:
Yeah seems like we want the second element here.

Reply:
Could also just do `state.boughtAmount = router.swapExactTokensForTokens(...)[1]`.

Reply:
Oops my bad. Looks like we want the second element (last element if more than 2 tokens in the path).